statsmodels/tsa/statespace/kalman_smoother.py

```diff
@@ -30,63 +31,117 @@
 ))
 
 
-def _kalman_smooth(model, kfilter, smoother_output):
-    # Create storage
-    scaled_smoothed_estimator = None
-    scaled_smoothed_estimator_cov = None
-    smoothing_error = None
-    smoothed_state = None
-    smoothed_state_cov = None
-    smoothed_state_disturbance = None
-    smoothed_state_disturbance_cov = None
-    smoothed_measurement_disturbance = None
-    smoothed_measurement_disturbance_cov = None
-
-    # Intermediate values
-    tmp_L = np.zeros((model.k_states, model.k_states, model.nobs),
-                     dtype=kfilter.dtype)
-
-    if smoother_output & (SMOOTHER_STATE | SMOOTHER_DISTURBANCE):
-        scaled_smoothed_estimator = np.zeros((model.k_states, model.nobs + 1),
-                                             dtype=kfilter.dtype)
-        smoothing_error = np.zeros((model.k_endog, model.nobs),
-                                   dtype=kfilter.dtype)
-    if smoother_output & (SMOOTHER_STATE_COV | SMOOTHER_DISTURBANCE_COV):
-        scaled_smoothed_estimator_cov = (
-            np.zeros((model.k_states, model.k_states, model.nobs + 1),
-                     dtype=kfilter.dtype)
-        )
+class _kalman_smoother(object):
```

Comment:
It's the same sort of think as the difference between the `KalmanFilter` python wrapper class and the `_dKalmanFilter` (or `_zKalmanFilter`, etc.) cython computation class. The underscored versions perform the actual filtering / smoothing work but are not user-friendly, so the interface is through the non-underscored versions. The main reason this evolved was that in cython there had to be a separate cython class for each data type.

There's less of a need for a distinction when the computation class is in python, but I kept the structure similar to the way it's done for filtering to hopefully make it easier to understand.

Also, this will ease the transition to introducing a cython version of the smoother.